circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/eccvm/msm_builder.hpp

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#include <cstddef>
+
+#include "./eccvm_builder_types.hpp"
+
+namespace proof_system {
+
+template <typename Flavor> class ECCVMMSMMBuilder {
+  public:
+    using CycleGroup = typename Flavor::CycleGroup;
+    using FF = typename Flavor::FF;
+    using Element = typename CycleGroup::element;
+    using AffineElement = typename CycleGroup::affine_element;
+
+    static constexpr size_t ADDITIONS_PER_ROW = proof_system_eccvm::ADDITIONS_PER_ROW;
+    static constexpr size_t NUM_SCALAR_BITS = proof_system_eccvm::NUM_SCALAR_BITS;
+    static constexpr size_t WNAF_SLICE_BITS = proof_system_eccvm::WNAF_SLICE_BITS;
+
+    struct MSMState {
+        uint32_t pc = 0;
+        uint32_t msm_size = 0;
+        uint32_t msm_count = 0;
+        uint32_t msm_round = 0;
+        bool msm_transition = false;
+        bool q_add = false;
+        bool q_double = false;
+        bool q_skew = false;
+
+        struct AddState {
+            bool add = false;
+            int slice = 0;
+            AffineElement point{ 0, 0 };
+            FF lambda = 0;
+            FF collision_inverse = 0;
+        };
+        std::array<AddState, 4> add_state{ AddState{ false, 0, { 0, 0 }, 0, 0 },
+                                           AddState{ false, 0, { 0, 0 }, 0, 0 },
+                                           AddState{ false, 0, { 0, 0 }, 0, 0 },
+                                           AddState{ false, 0, { 0, 0 }, 0, 0 } };
+        FF accumulator_x = 0;
+        FF accumulator_y = 0;
+    };
+
+    static std::vector<MSMState> compute_msm_state(const std::vector<proof_system_eccvm::MSM<CycleGroup>>& msms,
```

Comment:
This function would be very hard to understand what you do here for a person who hasn't implemented EC arithmetic by themselves several times. You have some comments at the start that assume that the person understands what reading and writing mean in the context of point lookup table, but this information is not in the file.

Reply:
I've added a short description of what I mean by "lookup table". I've added a link to the eccvm specification as well. I think the backround ecc knowledge required to understand this code is too much to communicate in the comments.